src/oscar/apps/offer/conditions.py

```diff
@@ -70,9 +69,12 @@ def is_partially_satisfied(self, offer, basket):
     def get_upsell_message(self, offer, basket):
```

Comment:
This method used to previously always return a string - now it can return `None`. Will that not cause problems [here](https://github.com/django-oscar/django-oscar/blob/81ec4da73acb703c94ebf9214d4e54aad2e5721a/src/oscar/apps/basket/views.py#L121-L122), because now `None` will get passed to the template instead of a string?

Reply:
(If not, is it worth adding a test to verify that `BasketView.get_upsell_messages` doesn't include any `None` values?)

Reply:
As I can see returning `None` won't cause any problems - `None` will be returned only for already applied offers  (only applied offers can have `delta <= 0`).

Please see the code below (lines 87-88). For applied offers both `offer.is_condition_partially_satisfied(basket)` and `offer not in applied_offers` are `False`.
https://github.com/django-oscar/django-oscar/blob/81ec4da73acb703c94ebf9214d4e54aad2e5721a/src/oscar/apps/basket/views.py#L81-L93

Will try to add some test to ensure my assumption is correct.

Reply:
@solarissmoke

Added additional test in https://github.com/django-oscar/django-oscar/pull/3548/commits/de86d7f43371f58d919d9d23e682b9cef0d2bbc8. Please re-check this PR when you will have a chance. Thanks!